### PR TITLE
Always increment Stream Id on createStream

### LINKED ISSUE #12065: Throwing protocol error when http2 max concurrent streams is exceeded instead of resStream
### Expected behavior
I'm new to Http2, and was trying to implement some load shedding on our server by limiting the number of concurrent streams set. This is currently fixed at 100. When setting the maxConcurrentStream in the HttpToHttp2ConnectionHandler, I was expecting to get see the behaviour mentioned in the spec (https://httpwg.org/specs/rfc7540.html#rfc.section.8.1.4)  and get a RST_STREAM frame back to the client. 

### Actual behavior
Instead i am getting an error, Stream ID does not exist. This results in a connection error, causing a GOAWAY with "PROTOCOL_ERROR" and closes the whole connection instead of just the stream? I have not understood if I am doing anything wrong here.

Screenshot 2022-02-01 at 15 50 15

Screenshot 2022-02-01 at 15 50 30

Screenshot 2022-02-01 at 16 32 55

### Minimal yet complete reproducer code (or URL to code)

Server Configuration

```
public class Http2OnlyHandler extends ApplicationProtocolNegotiationHandler {
 @Override
    protected void configurePipeline(ChannelHandlerContext ctx, String protocol) throws Exception {
        if (ApplicationProtocolNames.HTTP_2.equals(protocol)) {
            configureRedisPipeline(ctx);
            return;
        }

        if (ApplicationProtocolNames.HTTP_1_1.equals(protocol)) {
            configureFallback(ctx);
            return;
        }

        throw new IllegalStateException("Unsupported protocol: " + protocol);
    }


  private void configureRedisPipeline(ChannelHandlerContext ctx) {
  
    DefaultHttp2Connection connection = new DefaultHttp2Connection(true);
     InboundHttp2ToHttpAdapter listener = new InboundHttp2ToHttpAdapterBuilder(connection)
        .propagateSettings(true).validateHttpHeaders(false)
        .maxContentLength(maxContentLength).build();
  
    HttpToHttp2ConnectionHandler connectionHandler = new HttpToHttp2ConnectionHandlerBuilder()
        .frameListener(tester)
        .initialSettings(new Http2Settings().maxConcurrentStreams(100))
        .connection(connection).build();
  
    ctx.pipeline().addLast(connectionHandler);
  }
}


 bootstrap.group(bossGroup, workerGroup)
                .channel(channelClazz)
                .childHandler(new ChannelInitializer<SocketChannel>() {
                    @Override
                    protected void initChannel(SocketChannel ch) {
                        ch.pipeline().addLast(sslCtx.newHandler(ch.alloc()),
                            new Http2OnlyHandler(),
                            etc...)
                    }
                 })
```

Client:
```
channel.eventLoop().submit(() -> {
            int streamId = this.streamId.addAndGet(2);

            req.headers().setInt(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), streamId);
            req.headers().set(HttpConversionUtil.ExtensionHeaderNames.SCHEME.text(), "https");
            req.headers().set(HttpHeaderNames.HOST, "server");
            req.headers().add(HttpHeaderNames.ACCEPT_ENCODING, HttpHeaderValues.GZIP);
            req.headers().add(HttpHeaderNames.ACCEPT_ENCODING, HttpHeaderValues.DEFLATE);

            pendingRequests.put(streamId, new PendingReq<>(responseTransformer, future, System.currentTimeMillis(), (timeoutMs > 0) ? System.currentTimeMillis() + timeoutMs : -1));
            channel.writeAndFlush(req, channel.voidPromise());
        });

}
```

### Netty version
4.1.70.Final

### JVM version (e.g. `java -version`)
java version "1.8.0_271"
Java(TM) SE Runtime Environment (build 1.8.0_271-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.271-b09, mixed mode)
### OS version (e.g. `uname -a`)
Darwin sklas-mac 19.6.0 Darwin Kernel Version 19.6.0: Sun Nov 14 19:58:51 PST 2021; root:xnu-6153.141.50~1/RELEASE_X86_64 x86_64

### PR DESCRIPTION
**Pull request by alecharmon**
Motivation:

When the creation of a stream causes an error for whatever reason we want to increment the next expected stream id.
ie: checkNewStreamAllowed raises an error which causes the headers frame to get rejected subsequently a data frame arrives and it throws a protocol error.

Modification:

Use a finally block so that we always increment the expected next stream id

Result:

Fixes #12065

### CODE REVIEW

**Comment by normanmaurer** [comment]
@alecharmon did you sign our ical yet ? https://netty.io/s/icla

**Comment by alecharmon** [comment]
@normanmaurer I did

**Review by hyperxpro — APPROVED** [review]

**Comment by normanmaurer** [comment]
Also @vietj and @idelpivnitskiy

**Review by normanmaurer — APPROVED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 1]
```diff
@@ -748,13 +748,18 @@ public boolean canOpenStream() {
         public DefaultStream createStream(int streamId, boolean halfClosed) throws Http2Exception {
             State state = activeState(streamId, IDLE, isLocal(), halfClosed);
 
-            checkNewStreamAllowed(streamId, state);
+            try {
+                checkNewStreamAllowed(streamId, state);
+            } catch (Http2Exception e) {
+                throw e;
+            } finally {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java:755]: There is no need to rethrow. 
```suggestion
            } finally {
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java [hunk 2]
```diff
@@ -368,6 +368,21 @@ public void clientLocalIncrementAndGetStreamShouldRespectOverflow() throws Http2
...
+
+        assertThrows(Http2Exception.class, new Executable() {
+            @Override
+            public void execute() throws Throwable {
+                client.local().createStream(streamId, true);
+            }
+        });
+
+        int nextStreamID = client.local().incrementAndGetNextStreamId();
+        assertEquals(streamId, nextStreamID - 3);
```
Comment (idelpivnitskiy) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java:383]: It's hard to see how the change helps with the current test. Will be awesome if you can add a full integration test similar to the reproducer provided in #12065.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 1 above)
Comment (alecharmon) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Oh cool did not know that was the behavior but good to know - thanks @normanmaurer 

Updated ⚡

**Review by ejona86 — APPROVED** [review]

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java (hunk 2 above)
Comment (alecharmon) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java:383]: Spent most of my time today trying to create a new test in `codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ConnectionRoundtripTest.java` but having a hard time working with the api to recreate what the outcomes of https://github.com/netty/netty/issues/12065 are. Got any points/starting off points?

Apologies, first time working in this repo.

Comment (normanmaurer) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java:383]: seems like you did not push it yet ? Please do and I will have a look.

Comment (normanmaurer) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java:383]: let me merge this one first and I will work on a more complex test-case. The fix is correct

**Comment by normanmaurer** [comment]
@alecharmon thanks a lot!

**Pull request merged**